### ISSUE TITLE
Native vim support

### ISSUE THREAD

**Issue by paniash** [JuliaEditorSupport/julia-vim#222]
Hi!

I was wondering if there are plans to send this plugin to the official [Vim](https://github.com/vim/vim) repo. That would help ensure native support with runtime files instead of relying on a plugin. 

Thanks!

**Comment by paniash** [comment]
So, I opened an [issue](https://github.com/vim/vim/issues/7498) on the official vim repo. To the maintainers of this wonderful plugin, it'd be great to see this merged directly into vim. Thanks!

**Comment by paniash** [comment]
Any updates on this? @carlobaldassi

**Comment by clason** [comment]
Just to make clear: we're not talking about the full plugin here; just the syntax, indent, and ftplugin files. This requires no changes to _this_ plugin since you can override whatever is upstreamed (other filetype plugins to the same) -- all that is required of you is to give explicit permission for these files to be included in vim (and for @carlobaldassi to be listed as maintainer). 

(This would also be relevant for #269 since it would give (neo)vim users an option for a "light-weight" Julia editing experience if they don't need the L2U substitution.)

**Comment by clason** [comment]
@carlobaldassi If you do _not_ want to give permission, that is of course your prerogative, but then I'd appreciate it if you came out and said so -- then I'd drop this issue and move on with my life.

**Comment by carlobaldassi** [comment]
I have absolutely no problems with giving permission. Also, I certainly don't want to have people's lives hanging on it...
My real doubt is, in fact, about the L2U functionality, which is included in the plugins for the Julia language in the each and every other editor. Thus, I think it would be ideal if it worked the same in vim too. If it didn't, I wonder how many people will discover that the plugin for doing that exists.
Anyway, I guess that unless we can separately include the L2U plugin too into vim, that won't happen. This is why I haven't had the time to explore how to propose the inclusion in a way that makes sense yet.

**Comment by clason** [comment]
Sorry, I could have phrased this better -- the point is that I have time to do this now and follow up until it gets merged, but that will change later.

And I'm sorry if I gave the impression that I was down on the L2U functionality -- it's really nice and works smoothly (even on the commandline, which was surely not trivial to pull off!)

My proposal would be to just add the syntax highlighting and indentation files so that Julia files look nice out of the box. This is not hard, and I can do that both for vim and neovim. 

I do _not_ suggest including the L2U functionality; this essential part of editing Julia files would stay in this plugin. (If you're worried that this would make julia-vim "less attractive", I wouldn't worry -- people will still want to use it just for that!) I can't speak for vim, but it's definitely not a candidate for including in neovim...

I also do not suggest _removing_ the highlight files from `julia-vim` -- presumably, you want to keep improving them and adapting them to the newest versions of Julia. (Another reason to keep using the plugin ;)) If there are (big) enough changes, we can repeat the exercise of getting the latest versions into (neo)vim. This is pretty much how other language support in vim works, too (there's a full-featured language plugin, and the static "basics" are pulled in regularly).

To summarize, no change at all would be needed for julia-vim -- you'd just be "donating" the basic vim files to make (neo)vim work nicer out-of-the box.

**Comment by carlobaldassi** [comment]
Yes, I suppose that including a bare version of the plugin files is the only way for inclusion in vim, I don't really see any other way. So @clason if you want to take care of that, I'd be grateful. Otherwise, I'll do it myself when I have the time. After it's included, I also have no problem maintaining it whenever there are significant changes, of course.
(Aside: to clarify, I'm not worried that people won't use the plugin per se, I'm worried that people will consider vim support inferior to other editors. It's entirely possible that I'm overthinking it though. We'll see.)

**Comment by clason** [comment]
> Yes, I suppose that including a bare version of the plugin files is the only way for inclusion in vim, I don't really see any other way. So @clason if you want to take care of that, I'd be grateful.

I'd be happy to!

> After it's included, I also have no problem maintaining it whenever there are significant changes, of course.

You wouldn't have to maintain the vim files in addition; just this plugin like you've been doing. The point is just that if somebody opens an issue about Julia syntax highlighting in vim, Bram will say "please report the issue at julia-vim, and when it's fixed there, we'll just pull in their change".

> (Aside: to clarify, I'm not worried that people won't use the plugin per se, I'm worried that people will consider vim support inferior to other editors. It's entirely possible that I'm overthinking it though. We'll see.)

I think people value vim for different things than, say, VS Code, so I wouldn't worry. In any case, thank you for maintaining this plugin! Without your work, (neo)vim wouldn't be a viable Julia editor at all!

**Comment by clason** [comment]
Done: https://github.com/vim/vim/pull/8700

Filetype detection is already merged in https://github.com/vim/vim/commit/0eec8519424eaea7baebfda979c33dd609a4e3fa, syntax and indent files will be part of the next update of the runtime files.

**Comment by brammool** [comment]
My script to check runtime files finds a few problems:
The syntax file does not set and restore 'cpo', which is needed for the line continuation.  The other files do.

 Invalid expression: "'\U00-\U08' . '\U0A-\U1F'

Ftplugin file julia.vim: Vim(call):E117: Unknown function:
julia_blocks#remove_mappings in b:undo_ftplugin from julia.vim

**Comment by clason** [comment]
@brammool sorry to ask this, but maybe the second problem is an issue with your script? If I select these lines and `:'<,'>source` them, I get that error -- but when I source a script containing (only) these lines with `:source %`, it works fine?

**Comment by clason** [comment]
or is it just that the continuation operator `.` should be at the start of the new line instead of the end of the old one?

**Comment by clason** [comment]
@carlobaldassi my fixes are in https://github.com/clason/vim/commit/14a394a928dae34df8d6ae18730fabaf74e53e50 in case you want to apply them as well. (I can also make a PR when we're done if you prefer that.)

**Comment by clason** [comment]
Runtime files have been merged in vim and neovim now, so this can be closed now. Thanks for your help!

(For neovim, don't forget to `make distclean` before rebuilding!)

**Comment by paniash** [comment]
@clason Are the runtime files available only with the next release of neovim or is it possible to make it available for v0.5?

**Comment by clason** [comment]
They will not be backported, no, so if you don't want to use nightlies or build from master, you'll have to wait until 0.6.
